src/protocols/bdx/BdxMessages.h

```diff
@@ -0,0 +1,240 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines structures and utility methods for working with BDX
+ *      messages.
+ */
+
+#pragma once
+
+#include <support/BitFlags.h>
+#include <support/CodeUtils.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace BDX {
+
+enum TransferControlFlags : uint8_t
+{
+    // first 4 bits reserved for version
+    kSenderDrive   = 0x10,
+    kReceiverDrive = 0x20,
+    kAsync         = 0x40,
+};
+
+enum RangeControlFlags : uint8_t
+{
+    kDefLen      = 0x01,
+    kStartOffset = 0x02,
+    kWiderange   = 0x10,
+};
+
+struct TransferInit;
+using SendInit    = TransferInit;
+using ReceiveInit = TransferInit;
+
+/*
+ * A structure for representing a SendInit or ReceiveInit message (both contain
+ * identical parameters).
+ */
+struct TransferInit
+{
+    /**
+     * @brief
+     *  Pack (write) the message into a PacketBuffer.
+     *
+     * @param[out] aBuffer a PacketBuffer to use to write the message
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to write to the PacketBuffer
+     */
+    CHECK_RETURN_VALUE
+    CHIP_ERROR Pack(System::PacketBuffer & aBuffer) const;
+
+    /**
+     * @brief
+     *  Parse data from an PacketBuffer into a struct instance.
+     *
+     *  Note that this struct will store pointers that point into the passed PacketBuffer,
+     *  so it is essential that the PacketBuffer is not modified or freed until after this
+     *  struct is no longer needed.
+     *
+     * @param[in] aBuffer Pointer to a PacketBuffer containing the data.
+     *
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to read the message
+     */
+    CHECK_RETURN_VALUE
+    CHIP_ERROR Parse(const System::PacketBuffer & aBuffer);
+
+    /**
+     * @brief
+     *  Returns the size of buffer needed to write this message.
+     */
+    size_t PackedSize() const;
+
+    /**
+     * @brief
+     *  Equality check method.
+     */
+    bool operator==(const TransferInit &) const;
+
+    // Proposed Transfer Control (required)
+    BitFlags<uint8_t, TransferControlFlags> mTransferCtlFlags;
+    uint8_t mSupportedVersions = 0;
```

Comment:
This seems like s single version. Suggest adding a comment about how this is encoded and why an `s` on `Versions`.

Reply:
I think I may have misread the spec about this field. Is this correct -  for a transfer init message, this field will be filled in with a **single** proposed version. The response message will contain the chosen version for the transfer, which must be <= the proposed version? Do we ensure backward compatibility?

Reply:
The validation of the incoming version number is a protocol implementation detail. These data model encoding structures do not need to account for those rules. Agreed that it's a single version stated, and thus the name should be singular.